api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto

```diff
@@ -134,4 +134,10 @@ message TcpProxy {
   // payloads over a shared HTTP/2 tunnel. If this message is absent, the payload
   // will be proxied upstream as per usual.
   TunnelingConfig tunneling_config = 12;
+
+  // The maximum duration of a connection. The duration is defined as a period since a connection
+  // was established. If not set, there is no max duration. When max_connection_duration is reached
+  // the connection will be closed.
+  // Note: not implemented for upstream connections.
```

Comment:
Is it implemented for downstream connections? If not, would it make sense to build it as a single PR with both config changes and implementation?

Reply:
It does not implement yet. I will cover the implementation part also in this PR.

Reply:
I think if there were plans to do this for upstream, it'd be in cluster config.  Maybe just name it max_downstream_connection_duration and remove the note?

Reply:
I will do that. I think the name should be max_upstream_connection_duration.

Reply:
Hm, perhaps I'm confused about the goal here.  The PR description is "Added max_connection_duration for downstream connection" and the comment says it's not implemented for upstreams, so why would it be named max_upstream_connection_duration?

meanwhile if we want max upstream connection duration I'd think we'd want it in the connection pools.  Envoy's prefetching is still either alpha or not landed, but it's coming soon and if there's a max upstream duration I'd think we'd want it to be from connection establishment to connection teardown, not connection assignment to connection teardown.  WDYT?

Reply:
It's my bad, I will rename it. I thought that I have to add one more field like `// [#not-implemented-hide:]
  google.protobuf.Duration max_upstream_connection_duration = 14;
`. May I do also this?

Reply:
Is the one you have not sufficient?  Given that for the common case of TCP proxying, the lifetime of upstream connection is roughly that of downstream I'd hope what you have would be good enough.  If not I'd like to understand more about the use case, so I could understand if you actually want it to be max upstream (which would have to be in cluster config, not TCP proxy config) or max upstream-once-associated-with-downstream.

Reply:
I think it's sufficient. I am not planning to do this for upstream.